pkg/controller/statusmanager/pod_status.go

```diff
@@ -0,0 +1,278 @@
+type daemonsetState struct {
+	LastSeenStatus appsv1.DaemonSetStatus
+	LastChangeTime time.Time
+	Name           types.NamespacedName
+}
+
+// deploymentState is the same as daemonsetState.. but for deployments!
+type deploymentState struct {
+	LastSeenStatus appsv1.DeploymentStatus
+	LastChangeTime time.Time
+	Name           types.NamespacedName
+}
+
+// SetFromPods sets the operator Degraded/Progressing/Available status, based on
+// the current status of the manager's DaemonSets and Deployments.
+func (status *StatusManager) SetFromPods() {
+	status.Lock()
+	defer status.Unlock()
+
+	targetLevel := os.Getenv("RELEASE_VERSION")
+	reachedAvailableLevel := (len(status.daemonSets) + len(status.deployments)) > 0
+
+	progressing := []string{}
+	hung := []string{}
+
+	lastState := status.getLastPodState()
+
+	for _, dsName := range status.daemonSets {
+		ds := &appsv1.DaemonSet{}
+		if err := status.client.Get(context.TODO(), dsName, ds); err != nil {
+			log.Printf("Error getting DaemonSet %q: %v", dsName.String(), err)
+			progressing = append(progressing, fmt.Sprintf("Waiting for DaemonSet %q to be created", dsName.String()))
+			// Assume the OperConfig Controller is in the process of reconciling
+			// things; it will set a Degraded status if it fails.
+			continue
+		}
+
+		dsProgressing := false
+
+		if ds.Status.UpdatedNumberScheduled < ds.Status.DesiredNumberScheduled {
+			progressing = append(progressing, fmt.Sprintf("DaemonSet %q update is rolling out (%d out of %d updated)", dsName.String(), ds.Status.UpdatedNumberScheduled, ds.Status.DesiredNumberScheduled))
+			dsProgressing = true
+		} else if ds.Status.NumberUnavailable > 0 {
+			progressing = append(progressing, fmt.Sprintf("DaemonSet %q is not available (awaiting %d nodes)", dsName.String(), ds.Status.NumberUnavailable))
+			dsProgressing = true
+		} else if ds.Status.NumberAvailable == 0 { // NOTE: update this if we ever expect empty (unscheduled) daemonsets ~cdc
+			progressing = append(progressing, fmt.Sprintf("DaemonSet %q is not yet scheduled on any nodes", dsName.String()))
+			dsProgressing = true
+		} else if ds.Generation > ds.Status.ObservedGeneration {
+			progressing = append(progressing, fmt.Sprintf("DaemonSet %q update is being processed (generation %d, observed generation %d)", dsName.String(), ds.Generation, ds.Status.ObservedGeneration))
+			dsProgressing = true
+		}
+
+		if dsProgressing || ds.Annotations["release.openshift.io/version"] != targetLevel {
+			reachedAvailableLevel = false
+		}
+
+		dsState, ok := lastState.daemonsetState[dsName]
+		if ok && !reflect.DeepEqual(dsState.LastSeenStatus, ds.Status) {
+			ok = false // invalidate
+		}
+		if !ok {
+			dsState.Name = dsName
+			dsState.LastChangeTime = time.Now()
+			ds.Status.DeepCopyInto(&dsState.LastSeenStatus)
+			lastState.daemonsetState[dsName] = dsState
+		}
+
+		// Catch hung rollouts
+		if dsProgressing && ok && (time.Since(dsState.LastChangeTime)) > ProgressTimeout {
```

Comment:
You could move the `if dsProgressing` check out around the entire `lastState` thing.
Also, you should remove no-longer-`Progressing` DaemonSets from the state.